File: benchmark/quadrupedal_gaits.py
import sys
import time

import example_robot_data
import numpy as np

import crocoddyl
import pinocchio
from crocoddyl.utils.quadruped import SimpleQuadrupedalGaitProblem

T = int(5e3)  # number of trials
MAXITER = 1
WALKING = 'walk' in sys.argv
TROTTING = 'trot' in sys.argv
PACING = 'pace' in sys.argv
BOUNDING = 'bound' in sys.argv
JUMPING = 'jump' in sys.argv

GAIT = "walking"  # 104 nodes
if WALKING:
    print('running walking benchmark ...')
    GAIT = "walking"  # 104 nodes
if TROTTING:
    print('running trotting benchmark ...')
    GAIT = "trotting"  # 54 nodes
if PACING:
    print('running pacing benchmark ...')
    GAIT = "pacing"  # 54 nodes
if BOUNDING:
    print('running bounding benchmark ...')
    GAIT = "bounding"  # 54 nodes
if JUMPING:
    print('running jumping benchmark ...')
    GAIT = "jumping"  # 61 nodes


def runBenchmark(gait_phase):
    robot_model = example_robot_data.loadHyQ().model
    lfFoot, rfFoot, lhFoot, rhFoot = 'lf_foot', 'rf_foot', 'lh_foot', 'rh_foot'
    gait = SimpleQuadrupedalGaitProblem(robot_model, lfFoot, rfFoot, lhFoot, rhFoot)
    q0 = robot_model.referenceConfigurations['half_sitting'].copy()
    v0 = pinocchio.utils.zero(robot_model.nv)
    x0 = np.concatenate([q0, v0])

    type_of_gait = list(gait_phase.keys())[0]
    value = gait_phase[type_of_gait]
    if type_of_gait == 'walking':
        # Creating a walking problem
        ddp = crocoddyl.SolverDDP(
            gait.createWalkingProblem(x0, value['stepLength'], value['stepHeight'], value['timeStep'],
                                      value['stepKnots'], value['supportKnots']))
    elif type_of_gait == 'trotting':
        # Creating a trotting problem
        ddp = crocoddyl.SolverDDP(
            gait.createTrottingProblem(x0, value['stepLength'], value['stepHeight'], value['timeStep'],
                                       value['stepKnots'], value['supportKnots']))
    elif type_of_gait == 'pacing':
        # Creating a pacing problem
        ddp = crocoddyl.SolverDDP(
            gait.createPacingProblem(x0, value['stepLength'], value['stepHeight'], value['timeStep'],
                                     value['stepKnots'], value['supportKnots']))
    elif type_of_gait == 'bounding':
        # Creating a bounding problem
        ddp = crocoddyl.SolverDDP(
            gait.createBoundingProblem(x0, value['stepLength'], value['stepHeight'], value['timeStep'],
                                       value['stepKnots'], value['supportKnots']))
    elif type_of_gait == 'jumping':
        # Creating a jumping problem
        ddp = crocoddyl.SolverDDP(gait.createJumpingProblem(x0, value['jumpHeight'], value['timeStep']))

    duration = []
    xs = [robot_model.defaultState] * len(ddp.models())
    us = [m.quasicStatic(d, robot_model.defaultState) for m, d in list(zip(ddp.models(), ddp.datas()))[:-1]]
    for i in range(T):
        c_start = time.time()
        ddp.solve(xs, us, MAXITER, False, 0.1)
        c_end = time.time()
        duration.append(1e3 * (c_end - c_start))

    avrg_duration = sum(duration) / len(duration)
    min_duration = min(duration)
    max_duration = max(duration)
    return avrg_duration, min_duration, max_duration


# Setting up all tasks
if GAIT == 'walking':
    GAITPHASE = {
        'walking': {
            'stepLength': 0.25,
            'stepHeight': 0.25,
            'timeStep': 1e-2,
            'stepKnots': 25,
            'supportKnots': 2
        }
    }
elif GAIT == 'trotting':
    GAITPHASE = {
        'trotting': {
            'stepLength': 0.15,
            'stepHeight': 0.2,
            'timeStep': 1e-2,
            'stepKnots': 25,
            'supportKnots': 2
        }
    }
elif GAIT == 'pacing':
    GAITPHASE = {
        'pacing': {
            'stepLength': 0.15,
            'stepHeight': 0.2,
            'timeStep': 1e-2,
            'stepKnots': 25,
            'supportKnots': 2
        }
    }
elif GAIT == 'bounding':
    GAITPHASE = {
        'bounding': {
            'stepLength': 0.007,
            'stepHeight': 0.05,
            'timeStep': 1e-2,
            'stepKnots': 25,
            'supportKnots': 12
        }
    }
elif GAIT == 'jumping':
    GAITPHASE = {'jumping': {'jumpHeight': 0.5, 'timeStep': 1e-2}}

print('cpp-wrapped contact-forward dynamics on quadruped:')
avrg_duration, min_duration, max_duration = runBenchmark(GAITPHASE)
print('  CPU time [ms]: {0} ({1}, {2})'.format(avrg_duration, min_duration, max_duration))

File: benchmark/quadrupedal-gaits.cpp
///////////////////////////////////////////////////////////////////////////////
// BSD 3-Clause License
//
// Copyright (C) 2018-2019, LAAS-CNRS
// Copyright note valid unless otherwise stated in individual files.
// All rights reserved.
///////////////////////////////////////////////////////////////////////////////

#include <ctime>
#include <pinocchio/parsers/urdf.hpp>
#include <pinocchio/parsers/srdf.hpp>
#include "crocoddyl/multibody/utils/quadruped-gaits.hpp"
#include "crocoddyl/core/utils/callbacks.hpp"
#include "crocoddyl/core/solvers/fddp.hpp"

int main(int argc, char* argv[]) {
  bool CALLBACKS = false;
  unsigned int T = 5e3;  // number of trials
  unsigned int MAXITER = 1;
  if (argc > 1) {
    T = atoi(argv[1]);
  }

  pinocchio::Model rmodel;
  pinocchio::urdf::buildModel(HYQ_URDF, pinocchio::JointModelFreeFlyer(), rmodel);
  pinocchio::srdf::loadReferenceConfigurations(rmodel, HYQ_SRDF, false);

  crocoddyl::SimpleQuadrupedGaitProblem gait(rmodel, "lf_foot", "rf_foot", "lh_foot", "rh_foot");

  const Eigen::VectorXd& x0 = gait.get_defaultState();

  // Walking gait_phase
  const double stepLength(0.25), stepHeight(0.25), timeStep(1e-2);
  const unsigned int stepKnots(25), supportKnots(2);

  // DDP Solver
  boost::shared_ptr<crocoddyl::ShootingProblem> problem =
      gait.createWalkingProblem(x0, stepLength, stepHeight, timeStep, stepKnots, supportKnots);
  crocoddyl::SolverFDDP ddp(problem);
  if (CALLBACKS) {
    std::vector<boost::shared_ptr<crocoddyl::CallbackAbstract> > cbs;
    cbs.push_back(boost::make_shared<crocoddyl::CallbackVerbose>());
    ddp.setCallbacks(cbs);
  }

  // Initial State
  const std::size_t& N = ddp.get_problem()->get_T();
  std::vector<Eigen::VectorXd> xs(N + 1, x0);
  std::vector<Eigen::VectorXd> us(N, Eigen::VectorXd::Zero(problem->get_runningModels()[0]->get_nu()));
  for (std::size_t i = 0; i < N; ++i) {
    const boost::shared_ptr<crocoddyl::ActionModelAbstract>& model = problem->get_runningModels()[i];
    const boost::shared_ptr<crocoddyl::ActionDataAbstract>& data = problem->get_runningDatas()[i];
    model->quasiStatic(data, us[i], x0);
  }

  // Solving the optimal control problem
  struct timespec start, finish;
  double elapsed;
  Eigen::ArrayXd duration(T);
  for (unsigned int i = 0; i < T; ++i) {
    clock_gettime(CLOCK_MONOTONIC, &start);
    ddp.solve(xs, us, MAXITER, false, 0.1);
    clock_gettime(CLOCK_MONOTONIC, &finish);
    elapsed = static_cast<double>(finish.tv_sec - start.tv_sec) * 1000000;
    elapsed += static_cast<double>(finish.tv_nsec - start.tv_nsec) / 1000;
    duration[i] = elapsed / 1000.;
  }

  double avrg_duration = duration.mean();
  double min_duration = duration.minCoeff();
  double max_duration = duration.maxCoeff();
  std::cout << "  DDP.solve [ms]: " << avrg_duration << " (" << min_duration << "-" << max_duration << ")"
            << std::endl;

  // Running calc
  for (unsigned int i = 0; i < T; ++i) {
    clock_gettime(CLOCK_MONOTONIC, &start);
    problem->calc(xs, us);
    clock_gettime(CLOCK_MONOTONIC, &finish);
    elapsed = static_cast<double>(finish.tv_sec - start.tv_sec) * 1000000;
    elapsed += static_cast<double>(finish.tv_nsec - start.tv_nsec) / 1000;
    duration[i] = elapsed / 1000.;
  }

  avrg_duration = duration.sum() / T;
  min_duration = duration.minCoeff();
  max_duration = duration.maxCoeff();
  std::cout << "  ShootingProblem.calc [ms]: " << avrg_duration << " (" << min_duration << "-" << max_duration << ")"
            << std::endl;

  // Running calcDiff
  for (unsigned int i = 0; i < T; ++i) {
    clock_gettime(CLOCK_MONOTONIC, &start);
    problem->calcDiff(xs, us);
    clock_gettime(CLOCK_MONOTONIC, &finish);
    elapsed = static_cast<double>(finish.tv_sec - start.tv_sec) * 1000000;
    elapsed += static_cast<double>(finish.tv_nsec - start.tv_nsec) / 1000;
    duration[i] = elapsed / 1000.;
  }

  avrg_duration = duration.sum() / T;
  min_duration = duration.minCoeff();
  max_duration = duration.maxCoeff();
  std::cout << "  ShootingProblem.calcDiff [ms]: " << avrg_duration << " (" << min_duration << "-" << max_duration
            << ")" << std::endl;
}